k9mail/src/main/java/com/fsck/k9/controller/ImapSyncInteractor.java

```diff
@@ -0,0 +1,180 @@
+
+                if (Expunge.EXPUNGE_ON_POLL == account.getExpungePolicy()) {
+                    Timber.d("SYNC: Expunging folder %s:%s", account.getDescription(), folderName);
+                    imapFolder.expunge();
+                }
+                imapFolder.open(Folder.OPEN_MODE_RO);
+            }
+
+            notificationController.clearAuthenticationErrorNotification(account, true);
+
+            int remoteMessageCount = imapFolder.getMessageCount();
+            if (remoteMessageCount < 0) {
+                throw new IllegalStateException("Message count " + remoteMessageCount + " for folder " + folderName);
+            }
+
+            Timber.v("SYNC: Remote message count for folder %s is %d", folderName, remoteMessageCount);
+
+            int newMessages;
+            NonQresyncExtensionHandler handler = new NonQresyncExtensionHandler(syncHelper, flagSyncHelper,
+                    controller, messageDownloader);
+            newMessages = handler.continueSync(account, localFolder, imapFolder, listener);
+
+            int unreadMessageCount = localFolder.getUnreadMessageCount();
+            for (MessagingListener l : controller.getListeners()) {
+                l.folderStatusChanged(account, folderName, unreadMessageCount);
+            }
+
+            localFolder.setLastChecked(System.currentTimeMillis());
+            localFolder.setStatus(null);
+
+            Timber.d("Done synchronizing folder %s:%s @ %tc with %d new messages",
+                    account.getDescription(),
+                    folderName,
+                    System.currentTimeMillis(),
+                    newMessages);
+
+            for (MessagingListener l : controller.getListeners(listener)) {
+                l.synchronizeMailboxFinished(account, folderName, imapFolder.getMessageCount(), newMessages);
+            }
+
+            if (commandException != null) {
+                String rootMessage = MessagingController.getRootCauseMessage(commandException);
+                Timber.e("Root cause failure in %s:%s was '%s'",
+                        account.getDescription(), folderName, rootMessage);
+                localFolder.setStatus(rootMessage);
+                for (MessagingListener l : controller.getListeners(listener)) {
+                    l.synchronizeMailboxFailed(account, folderName, rootMessage);
+                }
+            }
+
+            Timber.i("Done synchronizing folder %s:%s", account.getDescription(), folderName);
+
+        } catch (AuthenticationFailedException e) {
+            controller.handleAuthenticationFailure(account, true);
+            for (MessagingListener l : controller.getListeners(listener)) {
+                l.synchronizeMailboxFailed(account, folderName, "Authentication failure");
+            }
+        } catch (Exception e) {
+            Timber.e(e, "synchronizeMailbox");
+            // If we don't set the last checked, it can try too often during
```

Comment:
This comment really wants moving to above line 135.

Or we could just test for the behaviour and describe the reason for the behaviour we want in the test name.

Reply:
Moved the comment